mdp/linear_flows_online.py

```diff
@@ -0,0 +1,658 @@
+
+    Here is a brief summary of the terminology used:
+       A "trainable" node: node.is_trainable() returns True, node.is_training() returns True.
+       A "trained" node: node.is_trainable() returns True, node.is_training() returns False.
+       A "non-trainalbe" node: node.is_trainable() returns False, node.is_training() returns False.
+
+    Differences between a Flow and an OnlineFlow:
+    a) In Flow, data is processed sequentially, training one node at a time. That is, the second
+       node's training starts only after the first node is "trained". Whereas, in an OnlineFlow data is
+       processed simultaneously training all the nodes at the same time.
+       Eg:
+
+       flow = Flow([node1, node2]), onlineflow = OnlineFlow([node1, node2])
+
+       Let input x = [x_0, x_1, ...., x_n], where x_t a sample or a mini batch of samples.
+
+       Flow training:
+            node1 trains on the entire x. While node1 is training, node2 is inactive.
+            node1 training completes. node2 training begins on the node1(x).
+
+            Therefore, Flow goes through all the data twice. Once for each node.
+
+       OnlineFlow training:
+            node1 trains on x_0. node2 trains on the output of node1 (node1(x_0))
+            node1 trains on x_1. node2 trains on the output of node1 (node1(x_1))
+            ....
+            node1 trains on x_n. node2 trains on the output of node1 (node1(x_n))
+
+            OnlineFlow goes through all the data only once.
+
+    b) Flow requires a list of dataiterables with a length equal to the
+       number of nodes or a single numpy array. OnlineFlow requires only one
+       input dataiterable as each node is trained simultaneously.
+
+    c) Additional train args (supervised labels etc) are passed to each node through the
+       node specific dataiterable. OnlineFlow requires the dataiterable to return a list
+       that contains tuples of args for each node: [x, (node0 args), (node1 args), ...]. See
+       train docstring.
+
+    d) OnlineFlow also has the cache attribute that retrieves any temporary caches variables
+       stored in the OnlineNodes.
+
+    Crash recovery is optionally available: in case of failure the current
+    state of the flow is saved for later inspection.
+
+    OnlineFlow objects are Python containers. Most of the builtin 'list'
+    methods are available. An 'OnlineFlow' can be saved or copied using the
+    corresponding 'save' and 'copy' methods.
+
+    """
+
+    def __init__(self, flow, crash_recovery=False, verbose=False):
+        super(OnlineFlow, self).__init__(flow, crash_recovery, verbose)
+        # check if the list of nodes is compatible.
+        self._check_compatibilitiy(flow)
+        # collect cache from individual nodes.
+        self._cache = self._get_cache_from_flow(flow)
+        # collect train_args for each node
+        self._train_arg_keys_list, self._train_args_needed_list = self._get_required_train_args_from_flow(flow)
```

Comment:
Yes it is needed and is checked for OnlineFlows too. OnlineFlow does not overwrite  "_check_nodes_consistency" method from its parent class.